lib/glific/providers/gupshup/partner/partner_api.ex

```diff
@@ -0,0 +1,148 @@
+      {:ok, %Tesla.Env{status: 200, body: body}} ->
+        {:ok, partner_token} =
+          Jason.decode!(body)
+          |> then(&Caches.set(@global_organization_id, "partner_token", &1["token"]))
+
+        {:ok, %{partner_token: partner_token}}
+
+      {_status, response} ->
+        {:error, "invalid response #{inspect(response)}"}
+    end
+  end
+
+  # fetches app id from phone using partner API
+  @spec get_apps_details(String.t()) :: {:ok, String.t()} | {:error, any}
+  defp get_apps_details(phone) do
+    with {:ok, app_list} <- Caches.get(@global_organization_id, "gupshup_apps"),
+         true <- app_list != false do
+      search_app(app_list, phone)
+    else
+      _ ->
+        do_get_apps_details(phone)
+    end
+  end
+
+  @spec do_get_apps_details(String.t()) :: {:ok, String.t()} | {:error, any}
+  defp do_get_apps_details(phone) do
+    url = @partner_url <> "/api/partnerApps"
+
+    with {:ok, %{partner_token: partner_token}} <- get_partner_token(),
+         {:ok, %Tesla.Env{status: 200, body: body}} = _response <-
+           get(url, headers: [{"token", partner_token}]) do
+      app_list =
+        body
+        |> Jason.decode!()
+        |> Map.get("partnerAppsList")
+
+      # Adding app list to cache
+      Enum.reduce(app_list, [], &(&2 ++ [%{"id" => &1["id"], "phone" => &1["phone"]}]))
+      |> then(&Caches.set(0, "gupshup_apps", &1))
+
+      search_app(app_list, phone)
+    end
+  end
+
+  @spec search_app(list(), String.t()) :: {:ok, String.t()} | {:error, any}
+  defp search_app(app_list, phone) do
+    app = Enum.find(app_list, fn app -> app["phone"] == phone end)
+
+    if app, do: {:ok, app["id"]}, else: {:error, "App not found"}
+  end
+
+  # fetches partner token first to get app access token
+  @spec get_app_token(String.t()) :: {:ok, map()} | {:error, any}
+  defp get_app_token(app_id) do
+    with {:ok, %{partner_token: partner_token}} <- get_partner_token() do
+      url = @app_url <> app_id <> "/token"
+
+      {:ok, app_token} = Caches.get(@global_organization_id, "app_token")
+
+      if app_token == false,
+        do: do_get_app_token(url, partner_token),
+        else: {:ok, %{app_token: app_token}}
+    end
+  end
+
+  @spec do_get_app_token(String.t(), String.t()) :: {:ok, map()} | {:error, any}
+  defp do_get_app_token(url, partner_token) do
+    get(url, headers: [{"token", partner_token}])
+    |> case do
+      {:ok, %Tesla.Env{status: 200, body: body}} ->
+        {:ok, app_token} =
+          Jason.decode!(body)
+          |> then(&Caches.set(@global_organization_id, "app_token", &1["token"]["token"]))
+
+        {:ok, %{app_token: app_token}}
+
+      {_status, _response} ->
+        {:error, "invalid credentials or app id}"}
+    end
+  end
+
+  @doc """
+  Fetches Partner token and App Access token to get tier information for an organization with input app id
+  """
+  @spec get_quality_rating(String.t()) :: {:error, any} | {:ok, map()}
+  def get_quality_rating(phone) do
```

Comment:
Curious why we are not using org_id and using the phone number here ?

Reply:
only phone number is needed in the API call